pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,960 @@
+// value types.
+type decoderType interface {
+	irType() ir.Type
+	encodeValueFields(
+		d *Decoder,
+		enc *jsontext.Encoder,
+		data []byte,
+	) error
+}
+
+// Type equivalent definitions
+type baseType ir.BaseType
+type pointerType ir.PointerType
+type structureType ir.StructureType
+type arrayType ir.ArrayType
+type voidPointerType ir.VoidPointerType
+type goSliceHeaderType ir.GoSliceHeaderType
+type goSliceDataType ir.GoSliceDataType
+type goStringHeaderType struct {
+	*ir.GoStringHeaderType
+	strFieldOffset uint32
+	strFieldSize   uint32
+	lenFieldOffset uint32
+	lenFieldSize   uint32
+}
+type goStringDataType ir.GoStringDataType
+type goMapType ir.GoMapType
+type goHMapHeaderType ir.GoHMapHeaderType
+type goHMapBucketType ir.GoHMapBucketType
+type goSwissMapHeaderType struct {
+	*ir.GoSwissMapHeaderType
+	// User-defined key and value type information
+	keyTypeID        ir.TypeID
+	keyTypeName      string
+	keyTypeSize      uint32
+	valueTypeID      ir.TypeID
+	valueTypeName    string
+	valueTypeSize    uint32
+	keyFieldOffset   uint32
+	valueFieldOffset uint32
+
+	// Internal Go swiss map representation fields
+	dirPtrOffset     uint32
+	dirPtrSize       uint32
+	dirLenOffset     uint32
+	dirLenSize       uint32
+	usedOffset       uint32
+	usedSize         uint32
+	ctrlOffset       uint32
+	ctrlSize         uint32
+	slotsOffset      uint32
+	slotsSize        uint32
+	groupFieldOffset uint32
+	groupFieldSize   uint32
+	dataFieldOffset  uint32
+	dataFieldSize    uint32
+	tableTypeID      ir.TypeID
+	groupTypeID      ir.TypeID
+	groupSliceTypeID ir.TypeID
+	elementTypeSize  uint32
+}
+type goSwissMapGroupsType ir.GoSwissMapGroupsType
+type goChannelType ir.GoChannelType
+type goEmptyInterfaceType ir.GoEmptyInterfaceType
+type goInterfaceType ir.GoInterfaceType
+type goSubroutineType ir.GoSubroutineType
+type eventRootType ir.EventRootType
+
+// Compile-time type assertions
+var (
+	_ decoderType = (*baseType)(nil)
+	_ decoderType = (*pointerType)(nil)
+	_ decoderType = (*structureType)(nil)
+	_ decoderType = (*arrayType)(nil)
+	_ decoderType = (*voidPointerType)(nil)
+	_ decoderType = (*goSliceHeaderType)(nil)
+	_ decoderType = (*goSliceDataType)(nil)
+	_ decoderType = (*goStringHeaderType)(nil)
+	_ decoderType = (*goStringDataType)(nil)
+	_ decoderType = (*goMapType)(nil)
+	_ decoderType = (*goHMapHeaderType)(nil)
+	_ decoderType = (*goHMapBucketType)(nil)
+	_ decoderType = (*goSwissMapGroupsType)(nil)
+	_ decoderType = (*goChannelType)(nil)
+	_ decoderType = (*goEmptyInterfaceType)(nil)
+	_ decoderType = (*goInterfaceType)(nil)
+	_ decoderType = (*goSubroutineType)(nil)
+	_ decoderType = (*eventRootType)(nil)
+)
+
+func newDecoderType(
+	irType ir.Type,
+	types map[ir.TypeID]ir.Type,
+) (decoderType, error) {
+	switch s := irType.(type) {
+	case *ir.GoSwissMapHeaderType:
+		dirPtrField, err := getFieldByName(s.RawFields, "dirPtr")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
```

Comment:
the repetition of the string `"malformed swiss map header type: "` makes me think you should extract a helper that returns a concrete type and an error and wrap it if it's non-nil...